### DISCUSSION TITLE
Next 9.3 getStaticPaths/getStaticProps and querystring only?

### DISCUSSION THREAD

**Discussion by PlopTheReal** [vercel/next.js#10951] [Help]
Is it possible to use getStaticPaths / getStaticProps with query string only? (single page app without any route)

**Comment by timneutkens** [comment]
I'm not entirely sure what you're asking for, but `getStaticPaths` generates static files, which means they can't access querystring at generation time, you can however access them during browser rendering using `useRouter`: https://nextjs.org/docs/api-reference/next/router#userouter

**Reply by PlopTheReal** [reply]
Thanks for the feedback. Let me explain in more details what I was looking for:
I got a single `pages/index.js` file. 
Website 'navigation' is like 
http://localhost/?p=welcome
http://localhost/?p=details
http://localhost/?p=aboutme

I've tried the following code but `context.query.p` is undefined so I had to fallback to `getServerSideProps`
```
export async function getStaticPaths() {
  return {
    paths: [
      { params: { p: "welcome" } },
      { params: { p: "details" } },
      { params: { p: "aboutme" } }
    ],
    fallback: false
  };
}

export async function getStaticProps(context) {
  let page = context.query.p
  let { items } = await fetchPageData(page)
  return { props: { page, items } }
}
```
Basically if I understand it properly `getStaticPaths` doesn't make a lot of sense for a single page app correct?

**Reply by timneutkens** [reply]
Think of `getStaticPaths` as generating HTML files, HTML files when served can't receive the querystring (unless you have a sophisticated proxy setup, in theory we could make this work on ZEIT Now, but we haven't as it wouldn't work when self-hosting)

General recommendation would be to use dynamic routes like `pages/[slug].js` and providing `slug`. Or when the pages are known just creating separate pages.

**Reply by PlopTheReal** [reply]
Ok thanks that makes more sense now. I'll stick with getServerSideProps for my usecase, that's fine anyway as fetched data is cached on the server.

**Reply by nihakue** [reply]
> Think of `getStaticPaths` as generating HTML files, HTML files when served can't receive the querystring (unless you have a sophisticated proxy setup, in theory we could make this work on ZEIT Now, but we haven't as it wouldn't work when self-hosting)
> 
> General recommendation would be to use dynamic routes like `pages/[slug].js` and providing `slug`. Or when the pages are known just creating separate pages.

Sorry to resurrect this issue, but with the addition of fallback=true and ISG, wouldn't it theoretically be possible to support query params for static pages without a proxy? I fail to see how dynamic routes and query params are any different in this regard, although of course I am only superficially familiar with the nextjs code.

**Reply by lucasbasquerotto** [reply]
Related: https://github.com/vercel/next.js/discussions/17269

**Reply by crevulus** [reply]
Forgive me for reopening this, but in the useRouter docs you linked (useRouter: https://nextjs.org/docs/api-reference/next/router#userouter) I don't see an implementation using ```getStaticPaths```. In fact, it says you can use it with ```getInitialProps```, ```getStaticProps```, and ```getServerSideProps```, but not ```Paths```. Using ```useRouter``` inside the ```getStaticPaths``` fn causes an invalid hook error.

What is the correct use if we want to access a query param in ```getStaticPaths```?

**Reply by ebenezerdon** [reply]
Hey @crevulus, did you find a work-around for accessing query params in `getStaticPaths`?
